### PR TITLE
KMA-51 - sync scores to AEM

### PR DESCRIPTION
**Pull request by wrandall22**
This PR creates a servlet for [scale-of-belief-lambda](https://github.com/CruGlobal/scale-of-belief-lambda) to access in order to save scores to AEM if they are entered through the scoring web interface.

This PR depends on https://github.com/CruGlobal/aem_osgi_config/pull/219

### CODE REVIEW

File: core/src/main/java/org/cru/contentscoring/core/util/SystemUtils.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+     * Retrieves the correct resource resolver for the subservice with according permissions.
+     *
+     * @param subservice the subservice the system user is registered with in User Mapper OSGI Config
+     * @return the ResourceResolver instance with correct system user permissions
+     * @throws LoginException if system user can't be retrieved correctly
+     */
+    ResourceResolver getResourceResolver(final String subservice) throws LoginException;
+
+    /**
+     * Gets the correct system user with permissions to replicate designation pages and replicates
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/util/SystemUtils.java]: 'designation pages' - copy/paste bug?

File: core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                resourcePathWithoutExtension,
+                parent,
+                resourceResolver,
+                homePagePath);
+
+            if (actualPath != null) {
+                updateScoreForPath(actualPath, resourceResolver, score);
+            }
+        }
+    }
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java]: This method is too big. Can you break it up please?

File: core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                Node node = contentResource.adaptTo(Node.class);
+                if (node != null) {
+                    node.setProperty("score", Integer.toString(score));
+
+                    Calendar now = Calendar.getInstance();
+                    ZonedDateTime zonedNow = ZonedDateTime.ofInstant(now.toInstant(), now.getTimeZone().toZoneId());
+
+                    node.setProperty(
+                        ContentScoreUpdateServiceImpl.CONTENT_SCORE_UPDATED,
+                        zonedNow.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java]: It looks like `ContentScoreUpdateServiceImpl` use a Calendar instance when setting this property (`CONTENT_SCORE_UPDATED`). Why are you using a formatted string here?

File: core/src/main/java/org/cru/contentscoring/core/service/SyncScoreService.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.ResourceResolver;
+
+import javax.jcr.RepositoryException;
+
+public interface SyncScoreService {
+    void syncScore(ResourceResolver resourceResolver, int score, String resourcePath, String resourceHost)
+        throws RepositoryException, ReplicationException, LoginException;
+}
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/SyncScoreService.java:18]: I think it'd be worth adding a sentence or two of docs to this interface (at the method or class level)

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        if (scoreIsValid(request.getParameter("score"))) {
+            score = Integer.valueOf(request.getParameter("score"));
+        } else {
+            response.sendError(400, "Invalid Score");
+            return;
+        }
+
+        final String resourcePath = request.getParameter("resourceUri[pathname]");
+        final String resourceHost = request.getParameter("resourceUri[hostname]");
+        final String resourceProtocol = request.getParameter("resourceUri[protocol]");
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: Why do you have the caller pass in these 3 uri components separately, instead of the caller passing a combined uri?

File: core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+    /**
+     * Checks if the provided vanity path is a valid redirect
+     * Modified from
+     * https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/wcm/vanity/impl/VanityURLServiceImpl.java
+     *
+     * @param pathScope The content path to scope the vanity path too.
+     * @param vanityPath Vanity path that needs to be validated.
+     * @param resourceResolver ResourceResolver object used for performing query/lookup
+     * @return return true if the vanityPath is a registered sling:vanityPath under /content
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java]: this paragraph seems outdated (the method returns String, not boolean, for instance)

File: core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+        if (parentMapResource != null) {
+
+            Resource protocolResource = parentMapResource.getChild(protocol);
+
+            if (protocolResource != null) {
+                String hostName = host;
+                if (homePage) {
+                    int lastIndexOfPeriod = host.lastIndexOf(".");
+                    hostName = host.substring(0, lastIndexOfPeriod) + "_" + host.substring(lastIndexOfPeriod + 1);
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java]: Is there a Sling api you can use to do this lookup for you? Seems like this is somewhat brittle.

File: core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+
+        return null;
+    }
+
+    private String determineSlingMappingPath() {
+        Map<String, String> environmentMap = ImmutableMap.of(
+            "local", "/etc/map.publish.local",
+            "dev", "/etc/map.publish.dev",
+            "uat", "/etc/map.publish.uat",
+            "prod", "/etc/map.publish.prod"
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImpl.java]: The values are identical, except for the suffix. Might be worth refactoring this. (Only relevant if you can't use a Sling api for mapping lookups, like I mentioned in my previous commit)

File: core/src/test/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImplTest.java [hunk 9]
```diff
@@ -300,16 +306,17 @@ private Hit mockHit(final String jcrPath) throws Exception {
...
+        when(queryBuilder.createQuery(any(PredicateGroup.class), eq(session))).thenReturn(query);
     }
 
-    private void mockSearch(final SimpleSearch search, final Hit... hits) throws Exception {
+    private void mockSearch(final Query query, final Hit... hits) {
         SearchResult searchResult = mock(SearchResult.class);
         when(searchResult.getHits()).thenReturn(Lists.newArrayList(hits));
-        when(search.getResult()).thenReturn(searchResult);
+        when(query.getResult()).thenReturn(searchResult);
+        when(query.getResult()).thenReturn(searchResult);
```
Comment (mattdrees) [inline core/src/test/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImplTest.java]: Why did you have to duplicate this line? The test succeeds when I delete the second line.

**Review by mattdrees — CHANGES_REQUESTED** [review]
It's been really hard to follow what's going on in SyncScoreServiceImpl, and it seems like what you are doing (replicating the sling url->content resolution algorithm) is bound to be brittle (for instance, I don't think your code would handle a redirect to another redirect). I don't think it's a good idea. Are there other options that you've explored?

I think the only reasonable thing to do is use Sling's resolution logic instead of replicating it. This might not be possible on the author jvm. 

For instance (and this might not be possible), set up a servlet that uses selectors instead of a /bin path, such that given a url like cru.org/foo/bar.html the servlet would respond at cru.org/foo/bar.scoring.json. And you would post the new score straight to that modified url. Sling would do its normal resource resolution, and the servlet would just need to grab the current resource.

That all assumes you can run an update servlet on the publisher, instead of the author, and that might not work. But I think we can work around that, if you think the concept is sound.

**Comment by mattdrees** [comment]
One (perhaps) useful thing about my proposal is that it has more potential for working with apache rewrite rules at the dispatcher, since the request would be traveling through the dispatcher.

**Comment by mattdrees** [comment]
Also, what Toggl project are you using for this?

**Comment by wrandall22** [comment]
The Toggl project is called `KMA-51`. It's under the K&M User Data Platform client.

File: core/src/test/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImplTest.java (hunk 9 above)
Comment (wrandall22) [reply core/src/test/java/org/cru/contentscoring/core/service/impl/SyncScoreServiceImplTest.java]: Probably came in by accident somehow.

**Comment by wrandall22** [comment]
I like the idea of using selectors; that would hugely reduce the code if we can get it to work. Tried a few tests locally, so far not getting it to hit my servlet.

**Comment by mattdrees** [comment]
Should be possible. Something like this: https://sling.apache.org/documentation/the-sling-engine/servlets.html#registering-a-servlet-using-java-annotations

**Comment by wrandall22** [comment]
Yes, definitely should be possible. Got most of the obvious changes out of the way locally already. I can push up a commit if you'd like to see (separate from this pr).

**Comment by wrandall22** [comment]
[Here's](https://github.com/CruGlobal/cru-aem-content-scoring/commit/f06af9788b6da060867e480a3d2169727f515c49) what I have so far.

**Comment by mattdrees** [comment]
Yeah, that looks like a good start. But POST requests aren't hitting your servlet?

**Comment by wrandall22** [comment]
Correct, I've tried with a few curl requests similar to:
`curl -X POST -H "Content-Type: application/json" --data @test-score-sync.json http://local.cru.org:4503/how-to-know-god.html.sync`

**Comment by mattdrees** [comment]
That should be `...sync.html`, I believe

**Comment by wrandall22** [comment]
Yeah, that's right. Though, it still doesn't hit my servlet :(

**Comment by wrandall22** [comment]
Same servlet gets hit if I go to the browser to this url (when adding a doGet()): http://local.cru.org:4503/how-to-know-god.sync.html

**Comment by wrandall22** [comment]
I think this must be conflicting with another servlet that is OOTB because if I post on author it creates that node name.

**Comment by mattdrees** [comment]
can you try using a different extension? maybe json instead of html

**Comment by wrandall22** [comment]
Yep, tried that too. Reading online about it.

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java [hunk 10]
```diff
@@ -49,14 +52,54 @@ protected void doPost(
...
+
+            // We're only scoring html pages
+            if (incomingUri.endsWith(".html")) {
+                incomingUri = incomingUri.replace(".html", "");
+                pathFinderResponse = client.target(incomingUri + ".find.path.html")
+                    .request()
+                    .get();
+            } else {
+                URI uri = new URI(incomingUri);
+                // This should be the load-balanced URL (e.g. https://www.cru.org), but could be a publisher URL.
```
Comment (wrandall22) [inline core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java:70]: This doesn't work, might need to use one of the publisher urls instead

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java [hunk 11]
```diff
@@ -49,14 +52,54 @@ protected void doPost(
...
+                    .toString();
+                pathFinderResponse = client.target(serverUri + "/bin/path/finder")
+                    .queryParam("path", incomingUri)
+                    .request()
+                    .get();
+            }
+
+            resourcePath = pathFinderResponse.readEntity(String.class);
+
+            if (resourcePath == null || !resourcePath.startsWith("/")) {
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: have you ever seen this call return something that didn't start with `/`?

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java [hunk 12]
```diff
@@ -73,7 +73,8 @@ protected void doPost(
                     .setHost(uri.getHost())
                     .build()
                     .toString();
-                pathFinderResponse = client.target(serverUri + "/bin/cru/path/finder")
+                LOG.debug("Calling {} with path {}", serverUri + "/bin/cru/path/finder.html", incomingUri);
+                pathFinderResponse = client.target(serverUri + "/bin/cru/path/finder.html")
```
Comment (mattdrees) [inline core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: I think it's a little weird to be returning a text blob from a .html url. Did you look into how hard it'd be to use a different extension, perhaps `.../finder.txt`?

**Review by mattdrees — APPROVED** [review]
Looking much more manageable, thanks.

Still a little bummed we have to do a search for vanity paths, but we can live with that i think.

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java (hunk 11 above)
Comment (wrandall22) [reply core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: Yes, I've seen it return html.

File: core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java (hunk 12 above)
Comment (wrandall22) [reply core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: Should be easy. Originally I was doing .html because the url was already coming in with it in many cases, but I ended up having to tack it on here manually anyway so it shouldn't be any harder to do a different extension.

Comment (mattdrees) [reply core/src/main/java/org/cru/contentscoring/core/servlets/SyncScoreServlet.java]: cool

**Review by mattdrees — APPROVED** [review]
PR looks good!
